Blacklist helper

PetrK39 opened the pull request:
Feel free to double-check it.
It works not as good as I expected with a few minutes of resource update delay but still better than copying in manually.
Ping me if docs need more explanation.

As the next step I would like to make some cleanup and then document all modules (maybe with some fancy 'jsdoc to md' github actions)

PetrK39 committed:
replace @include with @match
warned as being unsafe and obsolete in Manifest v3

PetrK39 committed:
Merge branch 'main' of https://github.com/PetrK39/gelbooru-overhaul-userscript

PetrK39 committed:
added hash function

PetrK39 committed:
created second userscript
specifically for updating blacklists
may require some tweaks after integration with the main script

PetrK39 committed:
no need to check for forbidden names

PetrK39 committed:
added isReadOnly and isUnremovable for default blacklists
added corresponding features for blacklist edit window

PetrK39 committed:
removed unnecessary no blacklists check

PetrK39 committed:
added loading blacklists from localStorage

PetrK39 committed:
added hash check on item updating

PetrK39 committed:
added timeout for loading from the localStorage

PetrK39 committed:
added typedef and readonly

PetrK39 committed:
documentation on uses and notes

Enchoseon commented:
As a heads up I'm going to be really delayed in responding for the next couple of weeks. Will check this as soon as I can.

Enchoseon commented on README.md:
```diff
@@ -8,6 +8,10 @@ Various toggleable changes to Gelbooru such as enlarging the gallery, removing t
 
 This script uses the Greasemonkey (`GM_*`) API. It has been developed/tested on [Tampermonkey](https://www.tampermonkey.net/) and [Violentmonkey](https://github.com/violentmonkey/violentmonkey).
 
+# Gelbooru Overhaul Blacklist Helper awailable now
```
```suggestion
# Gelbooru Overhaul Blacklist Helper Available Now
```
Fix typo

PetrK39 committed:
Update README.md

Co-authored-by: Encho <38841444+Enchoseon@users.noreply.github.com>

Enchoseon commented:
First run on a fresh install in a new browser in Violentmonkey 2.14.0 on Firefox 114.0.1, got this error:

```
index.php:544:1
[GELO]: Current page type is gallery Gelbooru Overhaul.user.js:209:25
Uncaught TypeError: scope.blacklistItems is undefined
    buildEditWindow moz-extension://77a38d0a-1ec2-4156-abd3-1911b458b984/ Gelbooru Overhaul.user.js#1:2725
    registerEditWinow moz-extension://77a38d0a-1ec2-4156-abd3-1911b458b984/ Gelbooru Overhaul.user.js#1:2655
    setupManager moz-extension://77a38d0a-1ec2-4156-abd3-1911b458b984/ Gelbooru Overhaul.user.js#1:3088
    applyTweakAdvancedBlacklist moz-extension://77a38d0a-1ec2-4156-abd3-1911b458b984/ Gelbooru Overhaul.user.js#1:1931
    onUpdate moz-extension://77a38d0a-1ec2-4156-abd3-1911b458b984/ Gelbooru Overhaul.user.js#1:732
    onUpdate moz-extension://77a38d0a-1ec2-4156-abd3-1911b458b984/ Gelbooru Overhaul.user.js#1:732
    applyConfig moz-extension://77a38d0a-1ec2-4156-abd3-1911b458b984/ Gelbooru Overhaul.user.js#1:789
    applyConfig moz-extension://77a38d0a-1ec2-4156-abd3-1911b458b984/ Gelbooru Overhaul.user.js#1:788
    applyConfig moz-extension://77a38d0a-1ec2-4156-abd3-1911b458b984/ Gelbooru Overhaul.user.js#1:787
    main moz-extension://77a38d0a-1ec2-4156-abd3-1911b458b984/ Gelbooru Overhaul.user.js#1:3503
    onDOMReady moz-extension://77a38d0a-1ec2-4156-abd3-1911b458b984/ Gelbooru Overhaul.user.js#1:171
index.php:2725:13
```

Crashes the entire script.

Here's the snippet the error points to:

```javascript
resources/gelbooru-overhaul.blacklistManager.js
221-            let nameList = document.createElement("datalist");
222-            nameList.setAttribute("id", "go-advBlacklistListNames");
223:            scope.blacklistItems.forEach(i => {
```

Still don't have time to look onto it properly, but posting this here for proper bookkeeping.

PetrK39 commented:
Looks like 'scope' is undefined. I'll try replace it with my recent discovery 'this.func = this.func.bind(this) when I get my hands on the code.
But I'm actually working on cleaning up code from things like that in a separate branch anyway

PetrK39 committed:
blacklist manager fixes
fixed blacklistItems is undefined
fixed error on applying OrderEntriesByHitCount while manager is not set up

PetrK39 committed:
Merge branch 'blacklist-helper' of https://github.com/PetrK39/gelbooru-overhaul-userscript into blacklist-helper

PetrK39 commented:
I just forgot to declare 'blacklistItems' in the class. Actually I got a different error but I believe this should fix it.
